File: app/backend/timesfm_studio/schemas/anomaly.py
"""Anomaly detection request and response schemas."""

from __future__ import annotations

from typing import Literal

from pydantic import BaseModel, ConfigDict, Field

from .dataset import ColumnMapping
from .forecast import ForecastConfigIn


class AnomalyRequest(BaseModel):
    model_config = ConfigDict(extra="forbid", frozen=True)

    dataset_id: str
    mapping: ColumnMapping
    horizon: int = Field(default=12, ge=1, le=512)
    critical_z: float = Field(default=3.0, gt=0)
    warning_z: float = Field(default=2.0, gt=0)
    forecast_config: ForecastConfigIn = Field(default_factory=ForecastConfigIn)


class ContextAnomalyRecord(BaseModel):
    date: str
    value: float
    trend: float
    residual: float
    z_score: float
    severity: Literal["CRITICAL", "WARNING", "NORMAL"]


class ForecastAnomalyRecord(BaseModel):
    date: str
    forecast: float
    q10: float
    q20: float
    q80: float
    q90: float
    severity: Literal["CRITICAL", "WARNING", "NORMAL"]


class AnomalySummary(BaseModel):
    total: int
    critical: int
    warning: int
    normal: int


class SeriesAnomalyResult(BaseModel):
    series_id: str
    res_std: float
    context_summary: AnomalySummary
    forecast_summary: AnomalySummary
    context_records: list[ContextAnomalyRecord]
    forecast_records: list[ForecastAnomalyRecord]


class AnomalyResponse(BaseModel):
    results: list[SeriesAnomalyResult]

File: app/frontend/src/types/anomaly.ts
import type { ColumnMapping } from"./dataset";

export type Severity = "CRITICAL" | "WARNING" | "NORMAL";

export interface ContextAnomalyRecord {
 date: string;
 value: number;
 trend: number;
 residual: number;
 z_score: number;
 severity: Severity;
}

export interface ForecastAnomalyRecord {
 date: string;
 forecast: number;
 q10: number;
 q20: number;
 q80: number;
 q90: number;
 severity: Severity;
}

export interface AnomalySummary {
 total: number;
 critical: number;
 warning: number;
 normal: number;
}

export interface SeriesAnomalyResult {
 series_id: string;
 res_std: number;
 context_summary: AnomalySummary;
 forecast_summary: AnomalySummary;
 context_records: ContextAnomalyRecord[];
 forecast_records: ForecastAnomalyRecord[];
}

export interface AnomalyResponse {
 results: SeriesAnomalyResult[];
}

export interface AnomalyRequest {
 dataset_id: string;
 mapping: ColumnMapping;
 horizon: number;
 critical_z: number;
 warning_z: number;
}
